T19028

tchx84 opened the pull request:
https://phabricator.endlessm.com/T19028

tchx84 committed:
basin-hatch: support videos

Handle video downloads and HTML document transfomations
to support video content from libingester2 hatches.

https://phabricator.endlessm.com/T19028

tchx84 committed:
packer: simplify basin manifest

There is no a real need for separating different types of
content in separate fields, other than content in general
and sets.

What we really need is a way of telling whether something
needs to be indexed (e.g, a real content) and when not to
(e.g. resources).

This gives us extra flexibility to combine things around,
e.g, having videos that are index and videos that are not.

https://phabricator.endlessm.com/T19028

tchx84 commented on tools/incubator.js:
```diff
@@ -32,6 +33,19 @@ const Incubator = new Lang.Class({
+
+        let session = new Soup.Session();
+        let message = Soup.Message.new('GET', uri);
+        let input_stream = session.send(message, null);
```
4myself: need to do something when this fails...

tchx84 commented on lib/basin-util.c:
```diff
@@ -95,6 +112,26 @@ basin_override_resources (const gchar *content)
+  objects = xmlXPathEvalExpression ("//a[@data-libingester-asset-id]", context);
+  if (!xmlXPathNodeSetIsEmpty (objects->nodesetval))
+    {
+      htmlDocPtr img_node = htmlReadDoc (THUMBNAIL_SVG, "", NULL, HTML_PARSE_RECOVER | HTML_PARSE_NOERROR);
```
4myself: I could do this in a similar but simpler way.

tchx84 commented:
@ptomato I added 2 comments for myself, things I forgot to add. But please go throught it.. I want your opinion on things like not actually extracting thumbnail and using a placeholder like I do now.

ptomato commented on lib/basin-util.c:
```diff
@@ -95,6 +112,26 @@ basin_override_resources (const gchar *content)
+          ekn_id = xmlGetProp (nodeset->nodeTab[i], "data-libingester-asset-id");
+          prop = g_strdup_printf ("ekn:///%s", ekn_id);
+          xmlSetProp (nodeset->nodeTab[i], "href", prop);
+          xmlFree (prop);
```
Should be `g_free()` — probably both call `free()` in practice, but always free API-allocated memory with the free function advised by that API

ptomato replied:
(And this is a pre-existing defect, but strictly speaking the elements of the list returned by `basin_find_resources()` should be freed with `xmlFree()`, but GJS has no way to know that ... so you should probably copy them with `g_strdup()` and free the original XML characters)

ptomato commented on tools/incubator.js:
```diff
@@ -32,6 +33,21 @@ const Incubator = new Lang.Class({
+    _download_file: function (uri) {
+        let session = new Soup.Session();
+        let message = Soup.Message.new('GET', uri);
+        let input_stream = session.send(message, null);
```
I think you can accomplish the same thing without Soup. You can just create a GFile for `uri`, `read()` it, and splice it to the temp file's output stream.

ptomato commented on examples/desc.json:
```diff
@@ -9,16 +9,15 @@
         "width": 540,
-        "source": "image.jpg"
+        "source": "image.jpg",
+        "notIndexed": true
```
`"indexed": false`? :smile:

ptomato reviewed (changes requested):
Hmm, the placeholder is a bit bad because there's nothing that the user can do even if they want to extract a thumbnail themselves. Maybe you could make the placeholder slightly more user friendly than a black SVG (e.g., a black SVG with a play icon) and provide an option in the manifest for the user to add a thumbnail if they are prepared to edit the manifest before using basin on it.

tchx84 committed:
basin-util: Use API-specific free() calls

https://phabricator.endlessm.com/T19028

tchx84 commented:
@ptomato we already provide the "play icon" svg, it comes from the article format CSS. It looks good IMHO.

![screenshot from 2017-09-07 11-25-13](https://user-images.githubusercontent.com/481418/30171209-40ad23b2-93bf-11e7-81af-13cac877ee6b.png)

ptomato commented on lib/basin-util.c:
```diff
@@ -32,6 +34,21 @@ basin_find_resources (const gchar *content)
+      for (int i = 0; i < nodeset->nodeNr; i++)
+        {
+          xmlChar *ekn_id = xmlGetProp (nodeset->nodeTab[i], "data-libingester-asset-id");
+          list = g_list_append (list, g_strdup (ekn_id));
```
Also a pre-existing problem, but I didn't spot it last time: `g_list_append()` is inefficient when adding multiple items, use `g_list_prepend()` and `g_list_reverse()` afterwards if the order matters.

tchx84 replied:
Thanks!

ptomato commented on tools/incubator.js:
```diff
@@ -85,6 +98,25 @@ const Incubator = new Lang.Class({
+        let file = this._download_file(asset['uri']);
+        let info = file.query_info('*', Gio.FileQueryInfoFlags.NONE, null);
+
+        basin_metadata['source'] = file.get_path();
```
Just realized another pre-existing defect that I didn't spot last time: you're setting these keys as properties on the `Map` object, and not as keys in the map's hashtable. To access the hashtable, you have to use `get()`, `set()`, and `has()`.

tchx84 replied:
ouch, I feel very silly. Can I fix this all-in-once in a posterior commit?

ptomato replied:
:+1:

ptomato commented on tools/incubator.js:
```diff
@@ -32,6 +32,19 @@ const Incubator = new Lang.Class({
     },
 
+    _download_file: function (uri) {
+        let source_file = Gio.File.new_for_uri(uri);
```
Hmm, maybe this was a bad suggestion on my part. I now get an "Operation not supported" exception thrown when I do this.

tchx84 replied:
Is this on endlessOS? I had that issue too, but it got fixed after installing glib-networking and gvfs on jhbuild.

ptomato replied:
I was using a flapjack runtime.

tchx84 replied:
I can just mimic the same behavior, with GFile.is_native and Soup, no big deal.

tchx84 committed:
basin-util: use prepend and revert

As per @ptomato suggestion, append is inenfficient and
is faster to prepend and revert.

https://phabricator.endlessm.com/T19028

tchx84 committed:
basin-incubator: drop Map() usage

We are not really using it properly and I find the current
syntax much easier to read than Map's sets and gets.

https://phabricator.endlessm.com/T19028

ptomato reviewed (approved):
OK. I didn't have time to test it in a flapjack sandbox this time, since my local flapjack is in a non-working state at the moment, but I won't block the PR on that.
